### PR TITLE
Implement a custom graph renaming routine for inline

### PR DESCRIPTION
**Pull request by jbachurski**
Addresses #82 by avoiding adding extraneous Identity nodes when inlining and supporting subgraph renaming

- Reimplements the `onnx.compose.add_prefix_graph` to make it more flexible.
- Slightly expands Spox scopes to support 'anonymous' reserved names, which don't have an associated `Var`/`Node`, but are nevertheless marked as taken. This is because not all inlined nodes/values have a respective `Var`/`Node`.

### CODE REVIEW

File: src/spox/_inline.py [hunk 1]
```diff
@@ -13,6 +15,48 @@
...
+    graph.CopyFrom(graph_)
+
+    for p in itertools.chain(graph.input, graph.initializer):
+        p.name = rename(p.name)
+    for si in graph.sparse_initializer:
+        si.values.name = rename(si.values.name)
+        si.indices.name = rename(si.indices.name)
+
+    for nd in graph.node:
+        if nd.name and rename_node is not None:
```
Comment (cbourjau) [inline src/spox/_inline.py:35]: Are nodes really allowed to have a missing `name` / `op_type`?

File: src/spox/_public.py [hunk 2]
```diff
@@ -218,13 +218,11 @@ def inline(model: onnx.ModelProto) -> _InlineCall:
     _signature_msg = f"signature {in_names}{_defaults_msg} -> {out_names}"
 
     model = _copy_model(model)
-    # FIXME: Renaming does not work on subgraphs as of ONNX 1.13/1.14.
     for node in model.graph.node:
         for attr in node.attribute:
-            if attr.HasField("g") or attr.graphs:
+            if attr.graphs:
```
Comment (cbourjau) [inline src/spox/_public.py:223]: Is there any technical reason why we cannot support this now?

**Review by cbourjau — APPROVED** [review]
Great! Thanks! I just have some minor suggestions/questions.

> This is because not all inlined nodes/values have a respective Var/Node.

Could you expand on this a bit? Do you mean the nodes inside a subgraph?

File: src/spox/_inline.py (hunk 1 above)
Comment (jbachurski) [reply src/spox/_inline.py:35]: https://github.com/onnx/onnx/blob/main/docs/IR.md#nodes

> name | string | An optional name of the node, used for diagnostic purposes only.

I think it's often left empty/unset in practice, though Spox does build the nodes with its internal naming.

You're right about `op_type`, though. FYI, replacing `op_type` is there for future-proofing as inlining models with functions might require renaming some local functions.

**Comment by jbachurski** [comment]
Looks like due to the conda ONNX version bump, the CI is now failing on opset generation, so #83 will fix this :)

**Comment by jbachurski** [comment]
I also added an explicit proto-based test, and one for subgraph lists as there is no practical example that can be constructed there (no node uses a list of subgraphs...). This is since we are now straying from the ONNX utility which would have some tests itself. Maybe I'll upstream the new one at some point?

> > This is because not all inlined nodes/values have a respective Var/Node.
> 
> Could you expand on this a bit? Do you mean the nodes inside a subgraph?

@cbourjau Yeah - say an inlined `ModelProto` has an input X and output Y, and an intermediate Z. Then there is a `Var` corresponding to X and another for Y, but there will be no explicit variable for Z in the Spox graph. However, it does have to reserve some name for itself in-scope to avoid e.g. other inlines of the same model using that name. So the `ScopeSpace` object marks whatever name was used for Z as taken with the newly applied patch.

**Comment by cbourjau** [comment]
Thanks for the clarification :)

**Pull request merged**